src/Marcher.hs

```diff
@@ -137,28 +137,50 @@ type Direction = Vec3
 
 -- | Marches a ray through a scene and then does shading, reflections and refractions.
 rayRender :: ImageSettings -> Scene -> Ray -> Color
-rayRender sett s ray = clamp $ case rayMarch sett s ray of
+rayRender sett s ray@(_, dir) = clamp $ case rayMarch sett s ray of
     Nothing -> getBackgroundColor sett
-    Just pos -> let color = getColor $ snd $ s pos
-              in case rayMarch sett newscene (pos + 3*epsilon `scale` calcNormal sett s pos, normalize $ light - pos) of
-        Nothing -> color
-        Just newpos -> if equalWithinError epsilon 0 $ mag (light-newpos) then color else gray * color
-
-  where light = getSunPosition sett
-        newscene = mergeScenes s $ pointToScene light
-        epsilon = getTolerance sett
+    Just pos -> let (dist, material) = s pos
+                    normal = calcNormal sett s pos
+                    light = getLight sett
+                in shade sett s material dir normal pos light
 
 -- | Marches a ray through a scene until it hits an object. Returns Nothing if it goes outside the scene.
 rayMarch :: ImageSettings -> Scene -> Ray -> Maybe Position
 rayMarch sett s (pos,dir)
     | end <= 0 = Nothing
-    | equalWithinError epsilon 0 dist = Just pos
+    | equalWithinError epsilon 0 dist = Just $ pos
     | otherwise = rayMarch sett{getRenderDistance=end-dist} s (pos + dist `scale` dir, dir) -- Each time lowering the distance to the end with the distance we traveled.
-    where   (dist, Material color _ _) = s pos
+    where   (dist, material) = s pos
             end = getRenderDistance sett
             epsilon = getTolerance sett
 
 
+shade :: ImageSettings -> Scene -> Material -> Direction -> Direction -> Position -> Light -> Color -- (materialColor, specPower, glossPower)
+shade sett s material eye normal pt (color, pos)
+    = clamp $ shadow `scale` (diffused + specular)
+    where   materialColor = getColor material -- intrinsic color
+            specPower = getSpecularLighting material -- specular property
+            glossPower = getGloss material -- gloss property
+            lightDir = normalize $ pos-pt -- direction to the light source
+            -- diffused stuff
+            lambert = max 0 (normal `dot` lightDir )
+            diffused = lambert `scale` (color*materialColor)
+            -- specular stuff
+            r = reflect normal lightDir
+            specular = max 0 (glossPower * (max 0 $ r `dot` eye)**specPower) `scale` color
+            -- neither of which apply in shadow
+            shadow  | lambert == 0 = 0
+                    | otherwise = case rayMarch sett{getRenderDistance=mag (pos-pt)} s (pt + 0.0001 `scale` normal, lightDir) of
```

Comment:
Wait, what is that 0.0001 doing, anyway?

Reply:
This is basically the same arbitrary value we had in the previous version of marching to the sun to compute the shadow.